### PR TITLE
Refactor binary operations

### LINKED ISSUE #5: Make result_dtype in _create_method functional
Pandas requieres the `result_dtype ` argument for the [_create_method](https://github.com/Julian-Harbeck/pandas-units-extension/blob/8c1f8b2b05b2e6e26e5e242f9c7816e0fae63c97/pandas_units_extension/units.py#L309) function of the `UnitsExtensionArray` class. This was added in 8c1f8b2b05b2e6e26e5e242f9c7816e0fae63c97, but it is currently not further used.

Clarify the following questions:
- Does `result_dtype` takes precedence over `coerce_to_dtype` argument?
- What if `result_dtype` is incompatible to the operator's return type?

Afterwards add implementation.

### PR DESCRIPTION
**Pull request by janpipek**
Note: I change the comparison logic so that unit-compatibility is always required, even for == and !=. It was confusing imho.

### CODE REVIEW

**Comment by Julian-Harbeck** [comment]
> Note: I change the comparison logic so that unit-compatibility is always required, even for == and !=. It was confusing imho.

It is a bit, but it also was compliant with astropy which would now change. See the following example:

```python
>>> 1 * u.m == 4 * u.s
False
>>> 1 * u.m != 4 * u.s
True
>>> 1 * u.m < 4 * u.s
astropy.units.errors.UnitConversionError: Can only apply 'less' function to quantities with compatible dimensions
```

**Comment by janpipek** [comment]
> > Note: I change the comparison logic so that unit-compatibility is always required, even for == and !=. It was confusing imho.
> 
> It is a bit, but it also was compliant with astropy which would now change. See the following example:
> 
> ```python
> >>> 1 * u.m == 4 * u.s
> False
> >>> 1 * u.m != 4 * u.s
> True
> >>> 1 * u.m < 4 * u.s
> astropy.units.errors.UnitConversionError: Can only apply 'less' function to quantities with compatible dimensions
> ```

You're right. Let me put it back. Pandas behaves similarly with incompatible types.

File: pandas_units_extension/units.py
Comment (Julian-Harbeck) [inline pandas_units_extension/units.py]: Maybe this would be a good time to move these lines up, directly under the definition of the `UnitsExtensionArray` class and before the accessor classes? Seems like a more resonable place in my opinion.

File: pandas_units_extension/units.py [hunk 1]
```diff
@@ -573,71 +574,49 @@ def isna(self):
...
             self_q: u.Quantity = as_quantity(self)
-            other_q: u.Quantity = as_quantity(other)
 
-            if is_comparison:
-                # Try apply conversion (we need same type for comparisons)
-                if is_array_like(other) and other.dtype != self.dtype:
-                    try:
-                        other_q = convert(other_q, self.unit)
-                    except InvalidUnitConversion:
-                        return _invalid_operator()
```
Comment (Julian-Harbeck) [inline pandas_units_extension/units.py:641]: I think by removing this we lose the special handling of temperature conversions from the `convert` function. On the other hand that might not be a bad thing? Because for arithmetic operations we are not doing that and it seems weird to only do it for comparison operations.

File: pandas_units_extension/units.py [hunk 2]
```diff
@@ -573,71 +574,49 @@ def isna(self):
...
 
-            elif is_scalar(other):
-                if is_comparison:
-                    return NotImplemented
+            self_q: u.Quantity = as_quantity(self)
+            try:
+                other_q = as_quantity(other)
+                result_q = op(self_q, other_q)
+            except (TypeError, UnitConversionError):
+                result_q = op(self_q, other)
```
Comment (Julian-Harbeck) [inline pandas_units_extension/units.py]: What is the motivation to try `op(self_q, other)` if we get a `TypeError` or `UnitConversionError` here?

File: tests/test_pandas_units_extension.py [hunk 3]
```diff
@@ -466,14 +475,80 @@ def test_temperature_comparison(self):
...
+        with pytest.raises(InvalidUnitConversion):
+            op(other, s1)
+
+    @pytest.mark.parametrize(
+        ("other", "result"),
+        [
+            pytest.param(1, pd.Series([False, False]), id="number"),
+            pytest.param("1 m", pd.Series([True, False]), id="string-as-unit"),
+            pytest.param("m", pd.Series([False, False]), id="string"),
+            pytest.param(1 * u.m, pd.Series([True, False]), id="unit"),
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: `1 * u.m` is not a unit but a Quantity, I would change the id therefore to `quantity`
```suggestion
            pytest.param(1 * u.m, pd.Series([True, False]), id="quantity"),
```

File: tests/test_pandas_units_extension.py
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: Add fixtures for equality and ordering comparison and use them throughout the `TestComparisonOps` class.
```suggestion
_all_equality_comparison_operators = [operator.eq, operator.ne]
_all_ordering_comparison_operators = [
    operator.le,
    operator.lt,
    operator.ge,
    operator.gt,
]


@pytest.fixture(
    params=_all_equality_comparison_operators + _all_ordering_comparison_operators
)
def comparison_op(request):
    return request.param


@pytest.fixture(params=_all_equality_comparison_operators)
def equality_comparison_op(request):
    return request.param


@pytest.fixture(params=_all_ordering_comparison_operators)
def ordering_comparison_op(request):
    return request.param
```

`equality_comparison_op` could be used in:
- `test_eq_ne`

`ordering_comparison_op` could be used in:
- `test_compare_scalar`
- `test_incomparable_units`
- `test_with_incompatible_non_units`

File: tests/test_pandas_units_extension.py [hunk 4]
```diff
@@ -426,6 +427,16 @@ def test_divmod(self, data):
...
 
+    @pytest.mark.xfail(reason="Will be resolved in a future PR.")
+    def test_radd(self):
+        result = (5 * u.cm) + pd.Series([1, 2, 3], dtype="unit[m]")
+        expected = pd.Series([105, 205, 305], dtype="unit[cm]")
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.xfail(reason="Makes no sense for pandas-provided fixtures")
+    def test_divmod_series_array(self, data, data_for_twos):
+        super().test_divmod_series_array(data, data_for_twos)
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: Why is that marked as `xfail`? For me it now `xpasses` which would make the overwrite obsolet?

File: tests/test_pandas_units_extension.py [hunk 5]
```diff
@@ -426,6 +427,16 @@ def test_divmod(self, data):
         self._check_divmod_op(ser, divmod, ser[0])
         self._check_divmod_op(ser[0], ops.rdivmod, ser)
 
+    @pytest.mark.xfail(reason="Will be resolved in a future PR.")
+    def test_radd(self):
+        result = (5 * u.cm) + pd.Series([1, 2, 3], dtype="unit[m]")
+        expected = pd.Series([105, 205, 305], dtype="unit[cm]")
+        tm.assert_series_equal(result, expected)
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: I will probably just add the remaining roperators that are currently commented out to `_all_arithmetic_operators` in the PR, then this test would be obsolet again, but we can keep it for now.

File: tests/test_pandas_units_extension.py
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: Why are the two equality operators removed?

If they stay removed could be simplified to
```suggestion
    @compare_scalar_mark_xfail
    def test_compare_scalar(self, data, ordering_comparison_op):
        return super().test_compare_scalar(data, ordering_comparison_op)
```

(The suggestion feature seems not to work when I mark lines where some lines in between have been deleted, the whole `@pytest.mark.parametrize` above can be substituted.

File: tests/test_pandas_units_extension.py [hunk 6]
```diff
@@ -466,14 +475,80 @@ def test_temperature_comparison(self):
...
+            operator.gt,
+        ],
+    )
+    def test_incomparable_units(self, op):
         s1 = pd.Series([1000, 2000, 3000], dtype="unit[m]")
         s2 = pd.Series([1000, 2000, 3000], dtype="unit[s]")
 
-        assert all(s1 != s2)
+        with pytest.raises(InvalidUnitConversion):
+            _ = op(s1, s2)
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: More of a design choice, but no real need to assign the result to `_`, the op works also without.
```suggestion
            op(s1, s2)
```

File: tests/test_pandas_units_extension.py [hunk 7]
```diff
@@ -466,14 +475,80 @@ def test_temperature_comparison(self):
...
+        "op",
+        [
+            operator.le,
+            operator.lt,
+            operator.ge,
+            operator.gt,
+        ],
+    )
+    def test_with_incompatible_non_units(self, op, other):
+        s1 = pd.Series([1000, 2000, 3000], dtype="unit[m]")
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: Everywhere else we just use `s` for a Series if there are not multiple of them in one test.

File: pandas_units_extension/units.py (hunk 1 above)
Comment (Julian-Harbeck) [reply pandas_units_extension/units.py:641]: Just saw it is used down below, but only for the path of the ordering comparison operators and not the one for `eq` and `ne`, should be the same for both?

File: pandas_units_extension/units.py [hunk 8]
```diff
@@ -573,71 +574,49 @@ def isna(self):
...
-            if is_equality:
-                return NotImplemented
-            else:
-                raise TypeError
+    def _create_arithmetic_method(cls, op):
+        op_name = f"__{op.__name__}__"
 
         def _binop(self, other):
             if isinstance(other, (ABCIndex, ABCSeries, ABCDataFrame)):
-                # rely on pandas to unbox and dispatch to us
```
Comment (Julian-Harbeck) [inline pandas_units_extension/units.py:615]: Why delete the comment here, but keep it in `_create_comparison_method._binop`? I tend to also keep the other comments like why we convert to Quantities and why divmod has to handled separately.

**Review by Julian-Harbeck — COMMENTED** [review]
Thanks @janpipek for giving this a look, trying to understand what happens in `_create_method` and `_binop` gave me quite a headache when I first started to understand what is going on. I therefore welcome the clearer `_create_arithmetic_method` and `_create_comparison_method` very much. Also the added tests help to define the indented interaction with various other elements.

However I think there still could be some unintended side effects and the tests got a bit confusing with the big parametrization sections, details see the comments below. For the tests I would recommend the usage of the new proposed fixtures `equality_comparison_op` and `ordering_comparison_op`.

File: tests/test_pandas_units_extension.py (hunk 4 above)
Comment (janpipek) [reply tests/test_pandas_units_extension.py]: Interesting :-) I must have fixed this by mistake...

File: tests/test_pandas_units_extension.py (hunk 5 above)
Comment (janpipek) [reply tests/test_pandas_units_extension.py]: Please do, when you have fixed the behaviour.

**Comment by Julian-Harbeck** [comment]
I think this PR also closes #5 as the `_create_method` is replaced by the higher up `_create_arithmetic_method` and `_create_comparison_method``.

File: pandas_units_extension/units.py (hunk 2 above)
Comment (janpipek) [reply pandas_units_extension/units.py]: I believe I met some edge case but cannot reproduce it now. I am deleting the code but be prepared if it resurfaces as bug.

File: tests/test_pandas_units_extension.py [hunk 9]
```diff
@@ -611,6 +672,8 @@ def test_unique(self):
         expected = UnitsExtensionArray([1, np.nan], unit="m")
         assert unique._unit == expected._unit
         np.testing.assert_equal(expected._value, unique._value)
+        assert unique._unit == expected._unit
+        np.testing.assert_equal(expected._value, unique._value)
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: These lines are doubled from merge from main.
```suggestion
```

File: tests/test_pandas_units_extension.py [hunk 10]
```diff
@@ -447,19 +458,9 @@ class TestComparisonOps(base.BaseComparisonOpsTests):
...
-            pytest.param(operator.lt, marks=compare_scalar_mark_xfail),
-            pytest.param(operator.ge, marks=compare_scalar_mark_xfail),
-            pytest.param(operator.gt, marks=compare_scalar_mark_xfail),
-        ],
-    )
-    def test_compare_scalar(self, data, comparison_op):
-        return super().test_compare_scalar(data, comparison_op)
+    @compare_scalar_mark_xfail
+    def test_compare_scalar(self, data, ordering_comparison_op):
+        return super().test_compare_scalar(data, ordering_comparison_op)
```
Comment (Julian-Harbeck) [inline tests/test_pandas_units_extension.py]: We are now not testing `test_compare_scalar` with `eq` and `ne`. One option would be to define another test for that:
```python
    def test_compare_scalar_equality(self, data, equality_comparison_op):
        return super().test_compare_scalar(data, equality_comparison_op)
```

Alternatively we can use an alternated version of how it was before, with `_all_equality_comparison_operators` always passing and `_all_ordering_comparison_operators` marked with the `compare_scalar_mark_xfail` marker:
```python
compare_scalar_mark_xfail: pytest.MarkDecorator = pytest.mark.xfail(
    pd.__version__ < "3.1.0",
    reason="Test fails on pandas below 3.1.0, see pandas GH #64365",
)


class TestComparisonOps(base.BaseComparisonOpsTests):
    @pytest.mark.parametrize(
        "op",
        [
            *_all_equality_comparison_operators,
            *[
                pytest.param(op, marks=compare_scalar_mark_xfail)
                for op in _all_ordering_comparison_operators
            ],
        ],
    )
    def test_compare_scalar(self, data, op):
        return super().test_compare_scalar(data, op)
```
The `compare_scalar_mark_xfail` marker then needs to be moved outside of the class apparently for pytest to discover it, alternatively skip the definition and just define it within the list comprehension.

File: pandas_units_extension/units.py [hunk 11]
```diff
@@ -572,85 +572,64 @@ def isna(self):
...
-                        other_q = convert(other_q, self._unit)
-                    except InvalidUnitConversion:
-                        return _invalid_operator()
+            if op_name in ["__eq__", "__ne__"]:
+                try:
+                    other_q: u.Quantity = as_quantity(other)
+                    if other_q.unit != self_q.unit:
+                        # This enables comparison of temperature values
+                        other_q = convert(other_q, self_q.unit)
+                    return op(self_q, other_q)
```
Comment (Julian-Harbeck) [inline pandas_units_extension/units.py:623]: This is now kind of duplicate with the path for the ordering comparison operators with the difference that `TypeError` and `InvalidUnitConversion` are not caught there. Maybe we could simplify this further like so:
```python
    @classmethod
    def _create_comparison_method(cls, op):
        op_name = f"__{op.__name__}__"

        def _binop(self, other):
            if isinstance(other, (ABCIndex, ABCSeries, ABCDataFrame)):
                # rely on pandas to unbox and dispatch to us
                return NotImplemented

            # Convert the thing to quantities
            self_q: u.Quantity = as_quantity(self)
            try:
                other_q: u.Quantity = as_quantity(other)
            except (TypeError, InvalidUnitConversion):
                # Compare things that cannot be converted to quantity, using astropy logic
                return op(self_q, other)

            if other_q.unit != self_q.unit:
                # This enables comparison of temperature values
                other_q = convert(other_q, self_q.unit)
            return op(self_q, other_q)

        return set_function_name(_binop, op_name, cls)
```

File: pandas_units_extension/units.py [hunk 12]
```diff
@@ -713,6 +692,11 @@ def value_counts(
         )
 
 
+UnitsExtensionArray._add_arithmetic_ops()
+UnitsExtensionArray._add_comparison_ops()
+UnitsExtensionArray.__pow__ = UnitsExtensionArray._create_method(operator.pow)
```
Comment (Julian-Harbeck) [inline pandas_units_extension/units.py]: With our overwrite of `_create_method` now removed this takes the base implementation path of `_create_method`. It seems to work as the tests for `pow` are passing, but I wonder if it should be changed to `_create_arithmetic_method` to use our implementation.

**Review by Julian-Harbeck — COMMENTED** [review]
@janpipek Thanks for addressing the first round of comments. I think we should not ignore `eq` and `ne` just because they successfully pass `test_compare_scalar` in comparison to the other four operators. Apart from that, the other three comments (accidently added one before starting the review) are more cosmetic changes, do as you like about them.

File: pandas_units_extension/units.py (hunk 11 above)
Comment (janpipek) [reply pandas_units_extension/units.py:623]: Not really - here "1 m" == "1 kg" will fail when the conversion does not work. But the convention says it should equal False. So we need to live without conversion for qe,ne. I will leave as it it, the duplicity does not seem so problematic.

Comment (Julian-Harbeck) [reply pandas_units_extension/units.py:623]: Good catch, thanks!
